solana/pyth2wormhole/program/src/types/mod.rs

```diff
@@ -431,39 +579,64 @@ mod tests {
     }
 
     #[test]
-    fn test_serialize_deserialize() -> Result<(), ErrBox> {
+    fn test_attestation_serde() -> Result<(), ErrBox> {
         let product_id_bytes = [21u8; 32];
         let price_id_bytes = [222u8; 32];
+        let attestation: PriceAttestation =
+            mock_attestation(Some(product_id_bytes), Some(price_id_bytes));
+
         println!("Hex product_id: {:02X?}", &product_id_bytes);
         println!("Hex price_id: {:02X?}", &price_id_bytes);
-        let attestation: PriceAttestation = PriceAttestation {
-            product_id: Pubkey::new_from_array(product_id_bytes),
-            price_id: Pubkey::new_from_array(price_id_bytes),
-            price: (0xdeadbeefdeadbabe as u64) as i64,
-            price_type: P2WPriceType::Price,
-            twap: P2WEma {
-                val: -42,
-                numer: 15,
-                denom: 37,
-            },
-            twac: P2WEma {
-                val: 42,
-                numer: 1111,
-                denom: 2222,
-            },
-            expo: -3,
-            status: P2WPriceStatus::Trading,
-            confidence_interval: 101,
-            corp_act: P2WCorpAction::NoCorpAct,
-            timestamp: 123456789i64,
-        };
 
         println!("Regular: {:#?}", &attestation);
         println!("Hex: {:#02X?}", &attestation);
-	let bytes = attestation.serialize();
+        let bytes = attestation.serialize();
         println!("Hex Bytes: {:02X?}", bytes);
 
-	assert_eq!(PriceAttestation::deserialize(bytes.as_slice())?, attestation);
+        assert_eq!(
+            PriceAttestation::deserialize(bytes.as_slice())?,
+            attestation
+        );
+        Ok(())
+    }
+
+    #[test]
+    fn test_attestation_serde_wrong_size() -> Result<(), ErrBox> {
+        assert!(PriceAttestation::deserialize(&[][..]).is_err());
+        assert!(PriceAttestation::deserialize(vec![0u8; 1].as_slice()).is_err());
+        Ok(())
+    }
+
+    #[test]
+    fn test_batch_serde() -> Result<(), ErrBox> {
+        let attestations: Vec<_> = (0..65535)
+            .map(|i| mock_attestation(Some([(i % 256) as u8; 32]), None))
+            .collect();
+
+        let serialized = batch_serialize(attestations.iter())?;
+
+        let deserialized = batch_deserialize(serialized.as_slice())?;
+
+        assert_eq!(attestations, deserialized);
+
+        Ok(())
+    }
+
+    #[test]
+    fn test_batch_serde_wrong_size() -> Result<(), ErrBox> {
+        assert!(batch_deserialize(&[][..]).is_err());
+        assert!(batch_deserialize(vec![0u8; 1].as_slice()).is_err());
+
+        let attestations: Vec<_> = (0..20)
+            .map(|i| mock_attestation(Some([(i % 256) as u8; 32]), None))
+            .collect();
+
+        let serialized = batch_serialize(attestations.iter())?;
+
+        // Missing last byte in last attestation must be an error
+        let len = serialized.len();
+        assert!(batch_deserialize(&serialized.as_slice()[..len - 1]).is_err());
+
```

Comment:
a good way to test serialization/deserialization is to write a randomized roundtrip test. Randomly generate a bunch of messages, then validate that each message round-trips through serialization and deserialization.

probably not necessary now, but good thing to think about in the future.